### PR TITLE
Style components using CSS modules

### PR DESCRIPTION
**Pull request by njdehoog**
This PR removes Tailwind CSS, which was used as a temporary styling solution. Instead, it introduces the use of [CSS modules](https://github.com/css-modules/css-modules) for styling. [Vite supports CSS modules out-of-the-box](https://vitejs.dev/guide/features#css-modules), so no additional dependencies are required for this.

The proposed way of working is to create a css module for every component in the component library, with the following file structure:
```
- /component-directory
  - index.js // The actual implementation of the component
  - style.module.css // The styles for the component
```

By using CSS modules, we ensure that styles are scoped to the module they are applied to, meaning that there is no chance of classname collisions.

## Example use
Importing a CSS modules file returns an object with a key for each classname. Here's a very basic example for a Button component:
```css
/* button-component/style.module.css */

.button {
  color: #FFF;
  background-color: #121212;
  height: 44px;
}
```

```js
/* button-component/index.js */

import styles from './style.module.css'

function Button({ text }) {
  return (<button className={styles.button}>{ text }</button>)
}
```

The HTML output will look something like this:

```html
<button class="_button_1gtnq_17">Click me</button>
```

## Sass support
CSS modules can also be Sass files, so `style.module.scss` works the same as `style.module.css`.

## CSS variables
This PR also adds CSS variables for shared styles, such as colours, spacing and fonts. Breakpoints, text size and spacing constants are imported from [@guardian/source-foundations](https://github.com/guardian/csnx/tree/main/libs/%40guardian/source-foundations).

CSS variables are partially generated based on `theme.config.js`. To generate new styles, run `pnpm generate:styles`. This will overwrite previously generated styles.

### CODE REVIEW

File: src/lib/components/molecules/table/index.jsx [hunk 1]
```diff
@@ -28,33 +30,42 @@ export function Table({ columns, data }) {
...
   return (
-    <table class="w-full table-fixed">
+    <table className={styles.table}>
       <thead>
         <tr>
           {table.columns.map((column, index) => (
-            <th key={column.id} className={column.headerCellClass}>
-              <HeaderCell key={index} onClick={() => sortByColumn(index)} {...column.headerProps} />
+            <th key={column.id} className={mergeStyles(styles.headerCell, column.styles?.headerCell)}>
+              <HeaderCell
```
Comment (ajwl) [inline src/lib/components/molecules/table/index.jsx:41]: I have to say - i find these three different lines relating to adding CSS quite confusing: 33, 36 and 40. Is there any way to remove the `mergeStyles()` parts. Or is that essential?

File: src/lib/components/molecules/table/table.stories.jsx [hunk 2]
```diff
@@ -61,21 +66,28 @@ export const Sortable = {
   },
 }
 
+const partyToColorMap = {
+  lab: 'lab',
+  con: 'con',
+  ld: 'libdem',
+  green: 'green',
+}
+
```
Comment (ajwl) [inline src/lib/components/molecules/table/table.stories.jsx:75]: do you think this kind of mapping helper function code has a place in this repo? If so perhaps we should pull this out into an easy to find shared helper file? (Doesn't have to be in this pull request).

File: src/lib/components/particles/stacked-bar/index.jsx [hunk 3]
```diff
@@ -16,22 +18,26 @@ export function StackedBar({ stack, width, height, createSVG = true }) {
     }
   }, [stack, width, height])
 
+  styles = mergeStyles({ ...defaultStyles }, styles)
+
   let totalWidth = 0
   const content = stack.map((d, index) => {
     const itemWidth = Math.round(d.fraction * width)
-    const textColor = isDarkColor(d.fill) ? 'fill-white' : 'fill-neutral-7'
+    const textColor = isDarkColor(d.fill) ? '#FFF' : '#121212'
```
Comment (ajwl) [inline src/lib/components/particles/stacked-bar/index.jsx:26]: can of course be added in a separate PR but can we use CSS variables here? eg  `var(--primary-text)` etc

File: src/lib/styles/foundation/partyColors.scss [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    --speaker: #707070;
+    --change: #494949;
+
+    --undeclared: #848484;
+  
+  }
+  
+/* BACKGROUND-COLOR -------------------------------------------- */ 
+
+.bg-color--con {
```
Comment (ajwl) [inline src/lib/styles/foundation/partyColors.scss:53]: If there's a better way of doing this - instead of all this repetition - i'm very open to it

Comment (njdehoog) [reply src/lib/styles/foundation/partyColors.scss:53]: I think we can generate it like some of the other styles. I'll pick that up at a later time though

File: src/lib/components/particles/stacked-bar/index.jsx (hunk 3 above)
Comment (njdehoog) [reply src/lib/components/particles/stacked-bar/index.jsx:26]: Yes, definitely. I think this depends on generating fill styles for all colours in our colour scheme though. I'll pick that up later.

File: src/lib/components/molecules/table/table.stories.jsx (hunk 2 above)
Comment (njdehoog) [reply src/lib/components/molecules/table/table.stories.jsx:75]: Good shout. The stories file is really just sample code, but I didn't need to use the mapping there at all. Will fix this by changing the data I pass to the table.

**Comment by ajwl** [comment]
Hello - this is a very impressive solution. I really like that the CSS files are very simple and clear, that the CSS is scoped to each component, and that we're not using any external libraries. 

However, the inclusion of the CSS in the JS is pretty complicated. Is there any way to simplify this? Could we just normal class names or omit the `mergeStyles` step? Or are these both integral.

Also v happy to just approve and we can iterate on this later

File: src/lib/components/molecules/table/index.jsx (hunk 1 above)
Comment (njdehoog) [reply src/lib/components/molecules/table/index.jsx:41]: Fair point. `mergeStyles()` is just a helper function that allows for overriding the styles, using the `styles` prop on the component or the column definition (see line 74 in `table.stories.jsx` for instance).

I agree that it would be better if we could automate this and remove all the boilerplate code. This is possible if we hook into the JSX compilation process, but that's somewhat complex, so I didn't want to take that on right away.

The choice was between this somewhat naïve solution for styled components, and using a library like [Emotion](https://emotion.sh/docs/introduction) or [Styled Components](https://styled-components.com/), which introduce a lot of new concepts. I felt it would be better for adoption and onboarding to stay as close as possible to how we currently write CSS. In future we should try and replicate some of the clever JSX magic that these libraries inject though.

That's my thinking, but I'm open to other suggestions.

File: src/lib/styles/helpers/mergeStyles.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  const merged = Object.keys(firstStyle).reduce((result, key) => {
+    let className = firstStyle[key]
+    if (secondStyle && key in secondStyle) {
+      className = className.concat(' ', secondStyle[key])
+    }
+    result[key] = className
+    return result
+  }, {})
+
+  return merged
```
[the remark this answers is not in the document]
Comment (njdehoog) [reply src/lib/styles/helpers/mergeStyles.js:19]: No all it's doing is joining class names together, so that styles can be overridden when using the component

**Comment by njdehoog** [comment]
Thanks for your review @ajwl. I agree that there is too much boilerplate right now, see earlier comment on `mergeStyles()`. I think if we can abstract that bit away then it becomes much easier to deal with.

I'm afraid that for CSS modules to work, the styles will always need to be imported and applied like that though. That's how the locally-scoped class names are applied. 

Having said that, in theory we could come up with some kind of naming convention and automagically apply the styles from the CSS module - definitely something to keep in mind.

Let me know if you're happy to approve with the changes I just pushed, on the basis that we will try and address these issues at a later time.

**Review by ajwl — APPROVED** [review]
Hello @njdehoog -thanks for detailed responses. I suppose the CSS imports will be set up at the start and from then on it's about just writing CSS which everyone is familiar with. 

Would be great to return to some of these merge style functions just to see if there is any more lightweight way of working here. Or simply just linking css to class names (tho I guess that would break the scoped nature of the styles?)

But happy to merge for now and then iterate later 

Thanks for your work and clear documentation of this
